File: Settings/src/com/android/tv/settings/device/apps/AppsFragment.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

package com.android.tv.settings.device.apps;

import static com.android.tv.settings.util.InstrumentationUtils.logEntrySelected;

import android.app.Activity;
import android.app.Application;
import android.app.tvsettings.TvSettingsEnums;
import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;

import androidx.annotation.Keep;
import androidx.annotation.Nullable;
import androidx.preference.Preference;

import com.android.settingslib.core.AbstractPreferenceController;
import com.android.tv.settings.PreferenceControllerFragment;
import com.android.tv.settings.R;
import com.android.tv.settings.overlay.FlavorUtils;
import com.android.tv.settings.util.SliceUtils;
import com.android.tv.twopanelsettings.slices.SlicePreference;

import java.util.ArrayList;
import java.util.List;

/**
 * Fragment for managing recent apps, and apps permissions.
 */
@Keep
public class AppsFragment extends PreferenceControllerFragment {

    private static final String KEY_PERMISSIONS = "Permissions";
    private static final String KEY_SECURITY = "security";
    private static final String KEY_PLAY_PROTECT = "play_protect";
    private static final String KEY_PLAY_AUTO_UPDATE = "play_auto_update";

    public static void prepareArgs(Bundle b, String volumeUuid, String volumeName) {
        b.putString(AppsActivity.EXTRA_VOLUME_UUID, volumeUuid);
        b.putString(AppsActivity.EXTRA_VOLUME_NAME, volumeName);
    }

    public static AppsFragment newInstance(String volumeUuid, String volumeName) {
        final Bundle b = new Bundle(2);
        prepareArgs(b, volumeUuid, volumeName);
        final AppsFragment f = new AppsFragment();
        f.setArguments(b);
        return f;
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        super.onCreatePreferences(savedInstanceState, rootKey);

        final Preference permissionPreference = findPreference(KEY_PERMISSIONS);
        final String volumeUuid = getArguments().getString(AppsActivity.EXTRA_VOLUME_UUID);
        permissionPreference.setVisible(TextUtils.isEmpty(volumeUuid));
        permissionPreference.setOnPreferenceClickListener(
                preference -> {
                    logEntrySelected(TvSettingsEnums.APPS_APP_PERMISSIONS);
                    return false;
                }
        );
        final Preference securityPreference = findPreference(KEY_SECURITY);
        final Preference playProtectPreference = findPreference(KEY_PLAY_PROTECT);
        final Preference playAutoUpdatePreference = findPreference(KEY_PLAY_AUTO_UPDATE);
        if (FlavorUtils.getFeatureFactory(getContext()).getBasicModeFeatureProvider()
                .isBasicMode(getContext())) {
            // playProtectPreference can be present only in two panel settings
            if (playProtectPreference != null) {
                playProtectPreference.setVisible(false);
            }
            if (isPlayProtectPreferenceEnabled(playProtectPreference)) {
                // By default show securityPreference unless playProtectPreference is enabled
                securityPreference.setVisible(false);
            }
        } else {
            if (isPlayProtectPreferenceEnabled(playProtectPreference)) {
                showPlayProtectPreference(playProtectPreference, securityPreference);
            } else {
                showSecurityPreference(securityPreference, playProtectPreference);
            }

            if (isPlayAutoUpdatePreferenceEnabled(playAutoUpdatePreference)) {
                playAutoUpdatePreference.setVisible(true);
            }
        }
    }

    private boolean isPlayProtectPreferenceEnabled(@Nullable Preference playProtectPreference) {
        return playProtectPreference instanceof SlicePreference
                && SliceUtils.isPlayTvSettingsSliceEnabled(
                        getContext(), ((SlicePreference) playProtectPreference).getUri());
    }

    private void showPlayProtectPreference(
            @Nullable Preference playProtectPreference,
            Preference securityPreference) {
        if (playProtectPreference != null) {
            playProtectPreference.setVisible(true);
        }
        securityPreference.setVisible(false);
    }

    private void showSecurityPreference(
            Preference securityPreference,
            @Nullable Preference playProtectPreference) {
        securityPreference.setVisible(true);
        if (playProtectPreference != null) {
            playProtectPreference.setVisible(false);
        }
    }

    private boolean isPlayAutoUpdatePreferenceEnabled(
            @Nullable Preference playAutoUpdatePreference) {
        return playAutoUpdatePreference instanceof SlicePreference
                && SliceUtils.isPlayTvSettingsSliceEnabled(
                        getContext(), ((SlicePreference) playAutoUpdatePreference).getUri());
    }

    @Override
    protected int getPreferenceScreenResId() {
        switch (FlavorUtils.getFlavor(getContext())) {
            case FlavorUtils.FLAVOR_TWO_PANEL:
            case FlavorUtils.FLAVOR_X:
            case FlavorUtils.FLAVOR_VENDOR:
                return R.xml.apps_x;
            default:
                return R.xml.apps;
        }
    }

    @Override
    protected List<AbstractPreferenceController> onCreatePreferenceControllers(Context context) {
        final Activity activity = getActivity();
        final Application app = activity != null ? activity.getApplication() : null;
        List<AbstractPreferenceController> controllers = new ArrayList<>();
        controllers.add(new RecentAppsPreferenceController(getContext(), app));
        return controllers;
    }

    @Override
    protected int getPageId() {
        return TvSettingsEnums.APPS;
    }
}

File: Settings/src/com/android/tv/settings/device/apps/AppsFragment.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */
package com.android.tv.settings.device.apps

import android.app.Activity
import android.app.tvsettings.TvSettingsEnums
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.provider.DeviceConfig
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Keep
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.Preference
import com.android.settingslib.core.AbstractPreferenceController
import com.android.tv.settings.PreferenceControllerFragment
import com.android.tv.settings.R
import com.android.tv.settings.library.util.LibUtils
import com.android.tv.settings.overlay.FlavorUtils
import com.android.tv.settings.util.InstrumentationUtils
import com.android.tv.settings.util.SliceUtilsKt
import com.android.tv.twopanelsettings.slices.SlicePreference
import kotlinx.coroutines.launch

/**
 * Fragment for managing recent apps, and apps permissions.
 */
@Keep
class AppsFragment : PreferenceControllerFragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        findPreference<Preference>(KEY_PERMISSIONS)?.apply {
            isVisible = TextUtils.isEmpty(
                    arguments?.getString(AppsActivity.EXTRA_VOLUME_UUID))
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                InstrumentationUtils.logEntrySelected(TvSettingsEnums.APPS_APP_PERMISSIONS)
                false
            }
        }
        findPreference<Preference>(KEY_HIBERNATED_APPS)?.isVisible = isHibernationEnabled

        findPreference<Preference>(KEY_ADD_APPS)?.apply {
            val addAppsUrl = requireContext().getString(R.string.add_apps_intent)
            if (addAppsUrl.isNotEmpty()) {
                isVisible = true
                intent = Intent.parseUri(addAppsUrl, 0)
            } else {
                isVisible = false
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                val securityPreference = findPreference<Preference>(KEY_SECURITY)
                val overlaySecuritySlicePreference =
                        findPreference<Preference>(KEY_OVERLAY_SECURITY)
                val updateSlicePreference = findPreference<Preference>(KEY_UPDATE)
                if (FlavorUtils.getFeatureFactory(requireContext()).getBasicModeFeatureProvider()
                                .isBasicMode(requireContext())) {
                    showSecurityPreference(securityPreference, overlaySecuritySlicePreference)
                    updateSlicePreference?.isVisible = false
                } else {
                    if (isOverlaySecuritySlicePreferenceEnabled(overlaySecuritySlicePreference)) {
                        showOverlaySecuritySlicePreference(
                                overlaySecuritySlicePreference, securityPreference)
                    } else {
                        showSecurityPreference(securityPreference, overlaySecuritySlicePreference)
                    }
                    updateSlicePreference?.isVisible =
                                isUpdateSlicePreferenceEnabled(updateSlicePreference)
                }
            }
        }
        return checkNotNull(super.onCreateView(inflater, container, savedInstanceState))
    }

    private suspend fun isOverlaySecuritySlicePreferenceEnabled(
            overlaySecuritySlicePreference: Preference?): Boolean {
        return (overlaySecuritySlicePreference is SlicePreference
                && SliceUtilsKt.isSettingsSliceEnabled(requireContext(),
                overlaySecuritySlicePreference.uri,
                TOP_LEVEL_SLICE_URI))
    }

    private fun showOverlaySecuritySlicePreference(
            overlaySecuritySlicePreference: Preference?,
            securityPreference: Preference?) {
        overlaySecuritySlicePreference?.isVisible = true
        securityPreference?.isVisible = false
    }

    private fun showSecurityPreference(
            securityPreference: Preference?,
            overlaySecuritySlicePreference: Preference?) {
        securityPreference?.isVisible = true
        overlaySecuritySlicePreference?.isVisible = false

    }

    private suspend fun isUpdateSlicePreferenceEnabled(
            updateSlicePreference: Preference?): Boolean {
        return (updateSlicePreference is SlicePreference
                && SliceUtilsKt.isSettingsSliceEnabled(
                requireContext(),
                updateSlicePreference.uri,
                TOP_LEVEL_SLICE_URI))
    }

    override fun getPreferenceScreenResId(): Int {
        return when (FlavorUtils.getFlavor(context)) {
            FlavorUtils.FLAVOR_X, FlavorUtils.FLAVOR_VENDOR -> R.xml.apps_x
            else -> R.xml.apps
        }
    }

    override fun onCreatePreferenceControllers(context: Context):
            List<AbstractPreferenceController> {
        val activity: Activity? = activity
        val app = activity?.application
        val controllers: MutableList<AbstractPreferenceController> = ArrayList()
        controllers.add(RecentAppsPreferenceController(getContext(), app))
        return controllers
    }

    override fun getPageId(): Int {
        return TvSettingsEnums.APPS
    }

    companion object {
        private const val KEY_PERMISSIONS = "Permissions"
        private const val KEY_SECURITY = "security"
        private const val KEY_OVERLAY_SECURITY = "overlay_security"
        private const val KEY_UPDATE = "update"
        private const val KEY_ADD_APPS="add_apps"
        private const val TOP_LEVEL_SLICE_URI = "top_level_settings_slice_uri"
        private const val KEY_HIBERNATED_APPS = "see_unused_apps"
        @JvmStatic
        fun prepareArgs(b: Bundle, volumeUuid: String?, volumeName: String?) {
            b.putString(AppsActivity.EXTRA_VOLUME_UUID, volumeUuid)
            b.putString(AppsActivity.EXTRA_VOLUME_NAME, volumeName)
        }

        fun newInstance(volumeUuid: String?, volumeName: String?): AppsFragment {
            val b = Bundle(2)
            prepareArgs(b, volumeUuid, volumeName)
            val f = AppsFragment()
            f.arguments = b
            return f
        }

        private val isHibernationEnabled: Boolean
            get() = DeviceConfig.getBoolean(
                    DeviceConfig.NAMESPACE_APP_HIBERNATION,
                    LibUtils.PROPERTY_APP_HIBERNATION_ENABLED, /* defaultValue= */ false)
    }
}